File: src/python/cameras/AlliedAlvium.py
"""
Allied Vision Alvium 1800 cameras for use with OpenCV.

Requires Allied Vision's Vimba for Python library (vmbpy). Download the
installation file vmbpy-*.whl. If you have installed the Vimba X Viewer
(https://www.alliedvision.com/de/produktportfolio/software/vimba-x-sdk/ ), the
file may already be present on your system. In my case it was located in
following directory:
    
    C:\Program Files\Allied Vision\Vimba X\api\python
    
For installation call pip with <x.y.z> being the Vimba version:

    pip install '<path>/vmbpy-<x.y.z>-py3-none-any.whl

@author: Marc Hensel
@contact: http://www.haw-hamburg.de/marc-hensel
@copyright: 2025
@version: 2025.03.06
@license: CC BY-NC-SA 4.0, see https://creativecommons.org/licenses/by-nc-sa/4.0/deed.en
"""

from Camera import Camera
from vmbpy import VmbSystem, PixelFormat, FrameStatus
import numpy as np
import cv2

class AlliedAlvium(Camera):
    """
    Tested cameras:    
        Alvium 1800 U-319 (2064×1544, 1/1.8", 54 fps)
        Alvium 1800 U-500 (2592×1944, 1/2.5", 68 fps)

    """

    # ========== Constructor ==================================================

    def __init__(self, camera_id=0, pixel_format='BGR8', bin_x=1, bin_y=1):
        """
        Initialize the camera.
        
        Parameters
        ----------
        camera_id : int, optional
            Camera id within all detected Vimba cameras. The default is 0.
        pixel_format : string, optional
            Pixel format as declared in Camera.pixel_formats. The default is 'BGR8'.
        bin_x : int, optional
            Number of pixels to bin horizontally (e.g., 2 will half the width). The default is 1.
        bin_y : int, optional
            Number of pixels to bin vertically (e.g., 2 will half the height). The default is 1.

        Returns
        -------
        None.

        """
        # Get Vimba system
        self.__vimba = VmbSystem.get_instance()
        self.__vimba.__enter__()
        
        # Find camera
        self.__camera = self.__get_camera(camera_id)
        self.__camera.__enter__()

        # Set pixel format (color or grayscale)
        assert pixel_format in Camera.pixel_formats
        self.__set_pixel_format('CV_BGR8' if pixel_format == 'BGR8' else 'CV_UINT8')

        # Set image size (binning)
        if bin_x != 1 or bin_y != 1:
            self.__set_binning(x=bin_x, y=bin_y)
        print(f'Sensor size  : {self.__camera.SensorWidth.get()} x {self.__camera.SensorHeight.get()} px')
        print(f'Image size   : {self.__camera.Width.get()} x {self.__camera.Height.get()} px')
        print(f'Frames / sec : {self.get_frame_rate()}')
        
        # Set acquisition parameters
        self.set_auto_exposure('Continuous')
        self.set_auto_white_balance('Continuous')
        self.set_auto_gain('Continuous')
                                        
        # Start acquisition
        print('Start streaming ... ', end="")
        self.__last_frame = None
        self.__handler = Handler()
        self.__camera.start_streaming(handler=self.__handler, buffer_count=10)
        while not self.__camera.is_streaming():
            cv2.waitKey(50)
        cv2.waitKey(500)
        print('done')

    # -------------------------------------------------------------------------

    def __get_camera(self, camera_id):
        """
        Get the Vimba camera.
        
        A camera ID can be specified for the case when more than one Vimba
        cameras are connected to the system. ID 0 corresponds to the first
        camera, ID 1 to the second, and so on.

        Parameters
        ----------
        camera_id : int
            Camera id within all detected Vimba cameras

        Raises
        ------
        Exception
            Raised if no camera was found.

        Returns
        -------
        camera : Camera
            Vimba camera that was found

        """
        cameras = self.__vimba.get_all_cameras()
        if len(cameras) > camera_id:
            camera = cameras[camera_id]
            print('Found camera : {}'.format(camera.get_name()))
        else:
            raise Exception('Alvium camera not found')
                
        return camera

    # -------------------------------------------------------------------------
    
    def __set_pixel_format(self, pixel_format):
        """
        Sets the pixel format of grabbed images.
        
        For use with OpenCV, following formats are supported:
            'CV_UINT8' :  8-bit gray (= color format pixel_format.Mono8)
            'CV_BGR8'  : 24-bit BGR (= color format pixel_format.Bgr8)
        
        Parameters
        ----------
        pixel_format : string
            Pixel format of grabbed frames. The default is color ('CV_BGR8').

        Returns
        -------
        None.

        """
        assert pixel_format in ['CV_UINT8', 'CV_BGR8']
        
        pixel_format = PixelFormat.Bgr8 if (pixel_format == 'CV_BGR8') else PixelFormat.Mono8
        supported_formats = self.__camera.get_pixel_formats()
        
        if pixel_format in supported_formats:
            self.__camera.set_pixel_format(pixel_format)
            print('Pixel format : {} (OpenCV)'.format(pixel_format))
        else:
            raise Exception('Pixel format {} (OpenCV) not supported'.format(pixel_format))

    # -------------------------------------------------------------------------
    
    def __set_binning(self, x=1, y=1):
        """
        Set the pixel binning.
        
        Binning N combines N sensor pixels to a single frame pixel. For
        instance, x = 2 halves the frame width, while y = 2
        halves the frame height.
        
        Note: Decreasing the binning, again, seems to have no effect until the
        camera has been reseted and detected, again. The reset command is
        called in the method release().

        Parameters
        ----------
        x : int, optional
            Number of pixels to bin horizontally. The default is 1.
        y : int, optional
            Number of pixels to bin vertically. The default is 1.

        Returns
        -------
        None.

        """
        self.__camera.BinningHorizontal.set(x)
        self.__camera.BinningVertical.set(y)

    # ========== Destructor ===================================================
    
    def release(self):
        """
        Reset camera.
        
        Reset is required, for instance, for binning. If horizontal or vertical
        binning has been applied to get smaller frames, the frame size will not
        increase, anymore, even when the binning values are set to 1. Calling
        the device reset solves this issue.

        Returns
        -------
        None.

        """
        print('Release camera : {}'.format(self.__camera.get_name()))
        self.__camera.stop_streaming()
        self.__camera.DeviceReset.run()
        self.__camera.__exit__(None, None, None)
        self.__vimba.__exit__(None, None, None)

    # ========== Frame grabbing ===============================================
    
    def get_frame(self):
        """
        Grab the last frame from the camera stream.

        Returns
        -------
        numpy.ndarray
            Frame in OpenCV compatible numpy format.

        """
        frame = self.__handler.frame
        if frame is None:
            print('Warning: No frame grabbed')
            return np.copy(self.__last_frame) if (self.__last_frame is not None) else None

        self.__last_frame = frame
        return np.copy(self.__last_frame)

    # ========== General properties ===========================================

    def get_name(self):
        """
        Get camera's name.

        Returns
        -------
        string
            Camera model name.

        """
        return self.__camera.get_name()

    # -------------------------------------------------------------------------

    def get_resolution(self):
        """
        Get width and height of grabbed frames.

        Returns
        -------
        int
            Width.
        int
            Height.

        """
        return int(self.__camera.Width.get()), int(self.__camera.Height.get())

    # -------------------------------------------------------------------------

    def set_resolution(self, width=None, height=None):
        """
        Set width and height of grabbed frames.

        Parameters
        ----------
        width : int, optional
            Width. The default is None.
        height : int, optional
            Height. The default is None.

        Returns
        -------
        None.

        """
        if width != None and height != None:
            self.__camera.stop_streaming()
            self.__camera.Width.set(width)
            self.__camera.Height.set(height)
            self.__camera.start_streaming(handler=self.__handler, buffer_count=10)

    # -------------------------------------------------------------------------
    
    def get_frame_rate(self):
        """
        Get the acquisition frame rate.

        Returns
        -------
        float
            Frames per second.

        """
        return self.__camera.AcquisitionFrameRate.get()

    # -------------------------------------------------------------------------

    def set_frame_rate(self, fps):
        """
        Set the acquisition frame rate.

        Parameters
        ----------
        fps : float
            Frames per second.

        Returns
        -------
        bool
            Returns True if the speed has been set, else False.

        """
        self.__camera.AcquisitionFrameRate.set(fps)
        return self.get_frame_rate() == fps


    # ========== Acquisition adjustments (image quality) ======================
    
    # ---------- Focus --------------------------------------------------------
    
    def set_autofocus(self, switch):
        """
        Enable/disable autofocus.

        This feature is not supported.

        """
        print('WARNING: Autofocus not supported')

    # ---------- Exposure time ------------------------------------------------

    def get_range_exposure_time_us(self):
        """
        Get parameter range of exposure time in microseconds.

        Returns
        -------
        float
            Minimum valid exposure time in [us].
        float
            Maximum valid exposure time in [us].

        """
        param_range = self.__camera.ExposureTime.get_range()
        return param_range[0], param_range[1]

    # -------------------------------------------------------------------------
        
    def set_exposure_time_us(self, time_us):
        """
        Set the exposure time in microseconds.

        Parameters
        ----------
        time_us : float
            Exposure time in [us].

        Returns
        -------
        bool
            True on success, else False.

        """
        # Get supported parameter range
        min_us, max_us = self.get_range_exposure_time_us()
        
        # Set parameter
        if min_us <= time_us <= max_us:
            self.set_auto_exposure('Off')
            self.__camera.ExposureTime.set(time_us)
            return self.__camera.ExposureTime.get() == time_us
        else:
            print(f'Warning: Exposure time not in range [{min_us}, {max_us}] us')
            return False

    # -------------------------------------------------------------------------

    def set_auto_exposure(self, mode):
        """
        Enable/disable auto exposure.

        Parameters
        ----------
        mode : string
            Mode as declared in Camera.modes.

        Returns
        -------
        None.

        """
        assert mode in Camera.modes
        self.__camera.ExposureAuto.set(mode)

    # ---------- Gain ---------------------------------------------------------
    
    def set_auto_gain(self, mode):
        """
        Enable/disable auto gain.

        Parameters
        ----------
        mode : string
            Mode as declared in Camera.modes.

        Returns
        -------
        None.

        """
        assert mode in Camera.modes
        self.__camera.GainAuto.set(mode)

    # ---------- White balance ------------------------------------------------

    def set_auto_white_balance(self, mode):
        """
        Enable/disable auto white balance.

        Parameters
        ----------
        mode : string
            Mode as declared in Camera.modes.

        Returns
        -------
        None.

        """
        assert mode in Camera.modes
        self.__camera.BalanceWhiteAuto.set(mode)

# ========== Handler ==========================================================

class Handler():
    """
    Class providing callback method for grabbed frames to the camera.
    
    """

    def __init__(self):
        self.frame = None

    # -------------------------------------------------------------------------
    
    def __call__(self, cam, stream, frame):
        if frame.get_status() == FrameStatus.Complete:
            self.frame = frame.as_opencv_image()      # TODO Clone image?
        cam.queue_frame(frame)

# -----------------------------------------------------------------------------
# Main (sample)
# -----------------------------------------------------------------------------

if __name__ == '__main__':
    camera = AlliedAlvium(bin_x=2, bin_y=2)
    camera.show_stream()
    camera.release()

File: src/c++/cameras/AlliedAlvium/AlliedAlvium.cpp
/*****************************************************************************************************
 * Allied Vision Alvium 1800 USB cameras for use with OpenCV.
 *****************************************************************************************************
 * Author: Marc Hensel, http://www.haw-hamburg.de/marc-hensel
 * Project: https://github.com/MarcOnTheMoon/imaging_learners/
 * Copyright: 2025, Marc Hensel
 * Version: 2025.03.15
 * License: CC BY-NC-SA 4.0, see https://creativecommons.org/licenses/by-nc-sa/4.0/deed.en
 *****************************************************************************************************/

// TODO: Wait for next frame in getFrame()?

/* Includes */
#include <mutex>
#include "AlliedAlvium.h"

/*****************************************************************************************************
 * Frame observer class
 *****************************************************************************************************/

/*! Constructor.
* 
* @param camera Camera to receive frames from
* @param cvPixelType OpenCV pixel type of frames (CV_8UC3 for BGR8, CV_8U for Mono8)
*/
FrameObserver::FrameObserver(CameraPtr camera, int cvPixelType) : IFrameObserver(camera) {
	this->cvPixelType = cvPixelType;
};

/*! Get next frame received from camera stream.
* 
* The method uses a mutex to prevent reading the frame while
* the callback is writing new data to it in parallel.
*
* @param frame Frame received from camera
* @return true if a non-empty frame was captured
*/
bool FrameObserver::getNextFrame(Mat& frame) {
	mtx.lock();
	bool isSuccess = isNewFrame;

	if (isNewFrame) {
		frame = this->frame.clone();
		isNewFrame = false;
	}
	mtx.unlock();
	return isSuccess;
}

/*! Callback for new frame received from the camera stream.
* 
* The method uses a mutex to prevent writing an OpenCV frame while
* data is read from the OpenCV frame object in parallel.
*
* @param rawFrame Grabbed frame
*/
void FrameObserver::FrameReceived(const FramePtr rawFrame) {
	// Convert to OpenCV format
	VmbUint32_t width, height;
	VmbUint8_t* data = NULL;
	rawFrame->GetWidth(width);
	rawFrame->GetHeight(height);
	rawFrame->GetImage(data);

	mtx.lock();
	frame = Mat((int)height, (int)width, cvPixelType, (void*)data).clone();
	isNewFrame = true;
	mtx.unlock();

	// Let object be filled with new grabbed frame
	m_pCamera->QueueFrame(rawFrame);
}

/*****************************************************************************************************
 * Constructor
 *****************************************************************************************************/

/*! Constructor.
* 
* @param cameraId Camera ID for all detected cameras of the given model. The default is 0.
*/
AlliedAlvium::AlliedAlvium(int cameraId, PixelFormat pixelFormat, int binX, int binY) {
	VmbSystem& vimba = VmbSystem::GetInstance();
	CameraPtrVector cameras;

	// Start Vimba API
	if (vimba.Startup() != VmbErrorSuccess) {
		throw exception("Error: Cannot start Vimba system");
	}

	// Find and open camera
	if ((vimba.GetCameras(cameras) == VmbErrorSuccess) && (cameras.size() >= cameraId)) {
		camera = cameras.at(cameraId);
		cout << "Found camera : " << getName() << endl;
	}
	else {
		throw exception("Error: Camera not found");
	}
	if (camera->Open(VmbAccessModeFull) != VmbErrorSuccess) {
		throw exception("Error: Cannot open camera");
	}

	// Set pixel format (color or grayscale)
	setPixelFormat(pixelFormat);
	cvPixelType = (pixelFormat == PixelFormat::BGR8) ? CV_8UC3 : CV_8U;

	// Set image size (binning)
	if ((binX != 1) || (binY != 1))
		setBinning(binX, binY);

	// Set acquisition parameters
	setAutoExposure(Mode::CONTINUOUS);
	setAutoGain(Mode::CONTINUOUS);
	setAutoWhiteBalance(Mode::ONCE);

	// Print properties (frame rate may depend on prior set parameters)
	Size resolution = getResolution();
	Size resolutionSensor = getResolutionSensor();
	cout << "Image size   : " << resolution.width << " x " << resolution.height << " px" << endl;
	cout << "Sensor size  : " << resolutionSensor.width << " x " << resolutionSensor.height << " px" << endl;
	cout << "Frame rate   : " << getFrameRate() << " fps" << endl;

	// Start acquisition
	cout << "Start streaming ... ";
	startImageAcquisition();
	cout << "done" << endl;
}

/*****************************************************************************************************/

/*! Set pixel format to BGR8 or Mono8.
*
* @param pixelFormat Pixel format (i.e., color or grayscale)
*/
bool AlliedAlvium::setPixelFormat(PixelFormat pixelFormat) {
	FeaturePtr feature;
	bool isError = false;

	// Set values
	String formatString = (pixelFormat == PixelFormat::BGR8) ? "BGR8" : "Mono8";
	isError = isError || (camera->GetFeatureByName("PixelFormat", feature) != VmbErrorSuccess);
	isError = isError || (feature->SetValue(formatString.c_str()) != VmbErrorSuccess);

	// Check success and return
	if (isError) {
		cout << "Warning: Cannot set pixel format" << endl;
	}
	return !isError;
}

/*****************************************************************************************************/

/*! Set the pixel binning.
* 
* Binning N combines N sensor pixels to a single frame pixel. For instance, x = 2 halves the frame width,
* while y = 2 halves the frame height.
* 
* Note: Decreasing the binning, again, seems to have no effect until the camera has been reseted and detected,
* again. The reset command is called in the method release().
* 
* @param binX Number of pixels to bin horizontally
* @param binY Number of pixels to bin vertically
* @return true on success, else false
*/
bool AlliedAlvium::setBinning(int binX, int binY) {
	FeaturePtr feature;
	bool isError = false;

	// Set values
	isError = isError || (camera->GetFeatureByName("BinningHorizontal", feature) != VmbErrorSuccess);
	isError = isError || (feature->SetValue(binX) != VmbErrorSuccess);
	isError = isError || (camera->GetFeatureByName("BinningVertical", feature) != VmbErrorSuccess);
	isError = isError || (feature->SetValue(binY) != VmbErrorSuccess);

	// Check success and return
	if (isError) {
		cout << "Warning: Cannot set binning" << endl;
	}
	return !isError;
}

/*****************************************************************************************************
 * Release camera
 *****************************************************************************************************/

 /*! Release and reset camera.
 */
void AlliedAlvium::release(void) {
	// Reset and close camera
	if (camera != (SharedPointer<VmbCPP::Camera>)NULL) {
		cout << "Release camera : " << getName() << endl;
		stopImageAcquisition();

		// Reset
		FeaturePtr feature;
		bool isError = false;
		isError = isError || (camera->GetFeatureByName("DeviceReset", feature) != VmbErrorSuccess);
		isError = isError || (feature->RunCommand() != VmbErrorSuccess);
		if (isError) {
			cout << "Warning: Cannot reset camera" << endl;
		}

		// Close
		camera->Close();
	}

	// Shutdown the Vimba API
	VmbSystem& vimba = VmbSystem::GetInstance();
	vimba.Shutdown();
}

/*****************************************************************************************************
 * Grab frame
 *****************************************************************************************************/

/*! Start asynchronous image acquisition.
* 
* Each new frame grabbed will invoke the callback method of the frame observer object.
*/
void AlliedAlvium::startImageAcquisition(void) {
	if (frameObserver == NULL) {
		// Create observer
		int numberFrameBuffers = 2;
		frameObserver = new FrameObserver(camera, cvPixelType);

		// Start acquisition
		if (camera->StartContinuousImageAcquisition(numberFrameBuffers, IFrameObserverPtr(frameObserver)) != VmbErrorSuccess) {
			cout << "Warning: Cannot start image acquisition" << endl;
		}
	}
}

/*! Stop asynchronous image acquisition.
*/
void AlliedAlvium::stopImageAcquisition(void) {
	if (frameObserver != NULL) {
		// Stop acquisition
		if (camera->StopContinuousImageAcquisition() != VmbErrorSuccess) {
			cout << "Warning: Cannot stop image acquisition" << endl;
		}

		// Observer (memory already deleted by StopContinuousImageAcquisition())
		frameObserver = NULL;
	}
}

/*! Get next frame from camera stream.
*
* @param frame Frame grabbed from camera
* @return true if a non-empty frame was captured
*/
bool AlliedAlvium::getFrame(Mat& frame) {
	return ((frameObserver != NULL) && frameObserver->getNextFrame(frame));
}

/*****************************************************************************************************
 * General properties
 *****************************************************************************************************/

 /*! Get camera model name.
 *
 * @return camera manufacturer and model name
 */
String AlliedAlvium::getName(void) {
	String name;
	camera->GetName(name);
	return name;
}

/*! Get width and height of grabbed frames.
*
* @return resolution containing width and height
*/
Size AlliedAlvium::getResolution(void) {
	FeaturePtr feature;
	VmbInt64_t width, height;
	bool isError = false;

	// Get values
	isError = isError || (camera->GetFeatureByName("Width", feature) != VmbErrorSuccess);
	isError = isError || (feature->GetValue(width) != VmbErrorSuccess);
	isError = isError || (camera->GetFeatureByName("Height", feature) != VmbErrorSuccess);
	isError = isError || (feature->GetValue(height) != VmbErrorSuccess);

	// Check success and return
	if (isError) {
		cout << "Warning: Cannot read resolution" << endl;
		return Size(0, 0);
	}
	return Size((int)width, (int)height);
}

/*! Set width and height of grabbed frames.
*
* @param width Image width to set.
* @param height Image height to set.
* @return true on success, else false
*/
bool AlliedAlvium::setResolution(int width, int height) {
	FeaturePtr feature;
	bool isError = false;

	// Set values
	stopImageAcquisition();
	isError = isError || (camera->GetFeatureByName("Width", feature) != VmbErrorSuccess);
	isError = isError || (feature->SetValue(width) != VmbErrorSuccess);
	isError = isError || (camera->GetFeatureByName("Height", feature) != VmbErrorSuccess);
	isError = isError || (feature->SetValue(height) != VmbErrorSuccess);
	startImageAcquisition();

	// Check success and return
	Size size = getResolution();
	isError = isError || (size.width != width) || (size.height != height);
	if (isError) {
		cout << "Warning: Cannot set resolution" << endl;
	}
	return !isError;
}

/*! Get width and height of the physical sensor.
*
* @return sensor resolution containing width and height
*/
Size AlliedAlvium::getResolutionSensor(void) {
	FeaturePtr feature;
	VmbInt64_t width, height;
	bool isError = false;

	// Get values
	isError = isError || (camera->GetFeatureByName("SensorWidth", feature) != VmbErrorSuccess);
	isError = isError || (feature->GetValue(width) != VmbErrorSuccess);
	isError = isError || (camera->GetFeatureByName("SensorHeight", feature) != VmbErrorSuccess);
	isError = isError || (feature->GetValue(height) != VmbErrorSuccess);

	// Check success and return
	if (isError) {
		cout << "Warning: Cannot read sensor resolution" << endl;
		return Size(0, 0);
	}
	return Size((int)width, (int)height);
}

/*! Get the acquisition frame rate.
*
* @return frames per second
*/
double AlliedAlvium::getFrameRate(void) {
	FeaturePtr feature;
	double fps;
	bool isError = false;

	// Get value
	isError = isError || (camera->GetFeatureByName("AcquisitionFrameRate", feature) != VmbErrorSuccess);
	isError = isError || (feature->GetValue(fps) != VmbErrorSuccess);

	// Check success and return
	if (isError) {
		cout << "Warning: Cannot read frame rate" << endl;
		return 0.0;
	}
	return fps;
}

/*! Set the acquisition frame rate.
*
* @param fps Frames per second
* @return true on success, else false
*/
bool AlliedAlvium::setFrameRate(double fps) {
	FeaturePtr feature;
	bool isError = false;

	// Set value (cannot be changed unless enabled)
	isError = isError || (camera->GetFeatureByName("AcquisitionFrameRateEnable", feature) != VmbErrorSuccess);
	isError = isError || (feature->SetValue(true) != VmbErrorSuccess);
	isError = isError || (camera->GetFeatureByName("AcquisitionFrameRate", feature) != VmbErrorSuccess);
	isError = isError || (feature->SetValue(fps) != VmbErrorSuccess);

	// Check success and return
	bool isError = abs(getFrameRate() - fps) > 1.0;
	if (isError) {
		cout << "Warning: Cannot set frame rate" << endl;
	}
	return !isError;
}

/*****************************************************************************************************
 * Acquisition adjustments (image quality)
 *****************************************************************************************************/

/*! Get Vimba X parameter string for mode.
* 
* @param mode Mode to get string for
* @return corresponding string to be used as parameter in Vimba X
*/
const char* AlliedAlvium::modeToString(Mode mode) {
	switch (mode) {
	case Mode::OFF:
		return "Off";
	case Mode::ONCE:
		return "Once";
	case Mode::CONTINUOUS:
		return "Continuous";
	default:
		throw exception("Warning: Invalid mode");
	}
}

/*! This feature is not supported.
*
* @return false
*/
bool AlliedAlvium::setAutofocus(Switch state) {
	cout << "Warning: Autofocus not supported" << endl;
	return false;
}

/*! Get parameter range of exposure time in microseconds.
*
* @param min [out] Minimum valid exposure time in [us]
* @param max [out] Maximum valid exposure time in [us]
* @return true on success, else false
*/
bool AlliedAlvium::getRangeExposureTimeMicroSecs(double* min, double* max) {
	FeaturePtr feature;
	bool isError = false;

	// Get values
	isError = isError || (camera->GetFeatureByName("ExposureTime", feature) != VmbErrorSuccess);
	isError = isError || (feature->GetRange(*min, *max) != VmbErrorSuccess);

	// Check success and return
	if (isError) {
		cout << "Warning: Cannot read range of exposure time" << endl;
	}
	return !isError;
}

/*! Set the exposure time in microseconds.
*
* @param exposureTime Exposure time in [us]
* @return true on success, else false
*/
bool AlliedAlvium::setExposureTimeMicroSecs(double exposureTime) {
	// Get supported parameter range
	double min, max;
	getRangeExposureTimeMicroSecs(&min, &max);

	// Set parameter
	if ((exposureTime >= min) && (exposureTime <= max)) {
		FeaturePtr feature;
		bool isError = false;

		// Set value
		isError = isError || !setAutoExposure(Mode::OFF);
		isError = isError || (camera->GetFeatureByName("ExposureTime", feature) != VmbErrorSuccess);
		isError = isError || (feature->SetValue(exposureTime) != VmbErrorSuccess);

		// Check success and return
		if (isError) {
			cout << "Warning: Cannot set exposure time" << endl;
		}
		return !isError;
	}
	else {
		cout << "Warning: Exposure time not in range [" << min << ", " << max << "] us" << endl;
		return false;
	}
}

/*! Enable/disable auto exposure.
*
* @param mode Target mode.
* @return true on success, else false
*/
bool AlliedAlvium::setAutoExposure(Mode mode) {
	FeaturePtr feature;
	bool isError = false;

	// Set value
	isError = isError || (camera->GetFeatureByName("ExposureAuto", feature) != VmbErrorSuccess);
	isError = isError || (feature->SetValue(modeToString(mode)) != VmbErrorSuccess);

	// Check success and return
	if (isError) {
		cout << "Warning: Cannot set auto exposure" << endl;
	}
	return !isError;
}

/*! Enable/disable auto gain.
*
* @param mode Target mode.
* @return true on success, else false
*/
bool AlliedAlvium::setAutoGain(Mode mode) {
	FeaturePtr feature;
	bool isError = false;

	// Set value
	isError = isError || (camera->GetFeatureByName("GainAuto", feature) != VmbErrorSuccess);
	isError = isError || (feature->SetValue(modeToString(mode)) != VmbErrorSuccess);

	// Check success and return
	if (isError) {
		cout << "Warning: Cannot set auto gain" << endl;
	}
	return !isError;
}

/*! Enable/disable auto white balance.
*
* @param mode Target mode.
* @return true on success, else false
*/
bool AlliedAlvium::setAutoWhiteBalance(Mode mode) {
	FeaturePtr feature;
	bool isError = false;

	// Set value
	isError = isError || (camera->GetFeatureByName("BalanceWhiteAuto", feature) != VmbErrorSuccess);
	isError = isError || (feature->SetValue(modeToString(mode)) != VmbErrorSuccess);

	// Check success and return
	if (isError) {
		cout << "Warning: Cannot set auto white balance" << endl;
	}
	return !isError;
}
